### PR TITLE
8262739: String inflation C2 intrinsic prevents insertion of anti-dependencies

### PR DESCRIPTION
**Pull request by TobiHartmann**
The string inflation intrinsic `StrInflatedCopyNode` which converts the String internal byte[] to char[] has `bottom_type` `Type::MEMORY`:
https://github.com/openjdk/jdk/blob/f71e8a61988cc9f095a4f034636baf04c3f48f77/src/hotspot/share/opto/intrinsicnode.hpp#L143-L145

Memory users (i.e. loads) should therefore be checked for nearby anti-dependent stores:
https://github.com/openjdk/jdk/blob/f71e8a61988cc9f095a4f034636baf04c3f48f77/src/hotspot/share/opto/node.cpp#L1482-L1486

However, since `InstructForm::captures_bottom_type` returns `false` for `StrInflatedCopyNode`, after matching, the corresponding `MachNode` version will have `bottom_type` `Type::BOTTOM` and no anti-dependency analysis will be performed for its users. Scheduling then incorrectly orders loads from the char array with anti-dependent stores to the same array, leading to incorrect results.

`StrCompressedCopyNode` does not need this because it has `TypeInt::INT` as  `bottom_type` (the memory output is handled via a `SCMemProjNode`):
https://github.com/openjdk/jdk/blob/f71e8a61988cc9f095a4f034636baf04c3f48f77/src/hotspot/share/opto/intrinsicnode.hpp#L131-L133

Thanks,
Tobias

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8262739](https://bugs.openjdk.java.net/browse/JDK-8262739): String inflation C2 intrinsic prevents insertion of anti-dependencies

### Reviewers
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**) ⚠️ Review applies to c132b1eb65ccea4f6c321a0afb8d71076359bb62
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/3025/head:pull/3025`
`$ git checkout pull/3025`

### CODE REVIEW

**Review by neliasso — APPROVED** [review]
Looks good.

**Comment by TobiHartmann** [comment]
Thanks for the review, Nils!

File: src/hotspot/share/adlc/formssel.cpp [hunk 1]
```diff
@@ -778,6 +778,7 @@ bool InstructForm::captures_bottom_type(FormDict &globals) const {
        !strcmp(_matrule->_rChild->_opType,"RotateLeft")   ||
        !strcmp(_matrule->_rChild->_opType,"RotateRight")   ||
 #if INCLUDE_SHENANDOAHGC
+       !strcmp(_matrule->_rChild->_opType,"StrInflatedCopy") ||
```
Comment (chhagedorn) [inline src/hotspot/share/adlc/formssel.cpp:781]: Should not be inside the ifdef.

**Review by chhagedorn — CHANGES_REQUESTED** [review]
Otherwise, the fix looks good!

File: src/hotspot/share/adlc/formssel.cpp (hunk 1 above)
Comment (TobiHartmann) [reply src/hotspot/share/adlc/formssel.cpp]: Yes, good catch, thank you! (That happens when doing last-minute re-ordering ...). Fixed.

**Review by chhagedorn — APPROVED** [review]

**Comment by TobiHartmann** [comment]
Thanks for the review, Christian!

**Pull request closed without merging**